pennylane/grouping/pauli.py

```diff
@@ -0,0 +1,254 @@
+    # in both Paulis.
+    if wire_map is None:
+        wire_labels = set(pauli_1.wires.labels + pauli_2.wires.labels)
+        wire_map = {label: i for i, label in enumerate(wire_labels)}
+
+    # Check if pauli_1 and pauli_2 are the same; if so, the result is the Identity
+    if are_identical_pauli_words(pauli_1, pauli_2):
+        first_wire = list(wire_map.keys())[0]
+        return Identity(first_wire)
+
+    # Compute binary symplectic representations
+    pauli_1_binary = pauli_to_binary(pauli_1, wire_map=wire_map)
+    pauli_2_binary = pauli_to_binary(pauli_2, wire_map=wire_map)
+
+    bin_symp_1 = np.array([int(x) for x in pauli_1_binary])
+    bin_symp_2 = np.array([int(x) for x in pauli_2_binary])
+
+    # Shorthand for bitwise XOR of numpy arrays
+    pauli_product = bin_symp_1 ^ bin_symp_2
+
+    return binary_to_pauli(pauli_product, wire_map=wire_map)
+
+
+def pauli_mult_with_phase(pauli_1, pauli_2, wire_map=None):
+    r"""Multiply two Pauli words together including the global phase.
+
+    Two Pauli operations can be multiplied together by taking the additive
+    OR of their binary symplectic representations. The phase is computed by
+    looking at the number of times we have the products  XY, YZ, or ZX (adds a
+    phase of :math:`i`), or YX, ZY, XZ (adds a phase of :math:`-i`).
+
+    Args:
+        pauli_1 (.Operation): A Pauli word.
+        pauli_2 (.Operation): A Pauli word to multiply with the first one.
+        wire_map  (dict[Union[str, int], int]): dictionary containing all wire labels used in the Pauli
+            word as keys, and unique integer labels as their values. If no wire map is
+            provided, the map will be constructed from the set of wires acted on
+            by the input Pauli words.
+
+    Returns:
+        tuple[.Operation, complex]: The product of ``pauli_1`` and ``pauli_2``, and the
+            global phase.
+
+    **Example**
+
+    This function works the same as ``pauli_mult`` but also returns the global
+    phase accumulated as a result of the Pauli product rules
+    :math:`\sigma_i \sigma_j = i \sigma_k`.
+
+    >>> from pennylane.pauli import pauli_mult_with_phase
+    >>> pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
+    >>> pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
+    >>> product, phase = pauli_mult_with_phase(pauli_1, pauli_2)
+    >>> product
+    PauliZ(wires=[0])
+    >>> phase
+    1j
+    """
+
+    # If no wire map is specified, generate one from the union of wires
+    # in both Paulis.
+    if wire_map is None:
+        wire_labels = set(pauli_1.wires.labels + pauli_2.wires.labels)
+        wire_map = {label: i for i, label in enumerate(wire_labels)}
+
+    # Get the product; use our earlier function
+    pauli_product = pauli_mult(pauli_1, pauli_2, wire_map)
+
+    pauli_1_names = [pauli_1.name] if isinstance(pauli_1.name, str) else pauli_1.name
+    pauli_2_names = [pauli_2.name] if isinstance(pauli_2.name, str) else pauli_2.name
```

Comment:
What does the if statement protect against?

Reply:
The case where there is a non-identity operation on just a single qubit, the name is a stored as a single string instead of a list. Will add a comment :+1: